src/posix/net.cpp

```diff
@@ -1463,6 +1463,8 @@ void CurlHttpIO::send_request(CurlHttpContext* httpctx)
         curl_easy_setopt(curl, CURLOPT_TCP_KEEPINTVL, 60L);
         curl_easy_setopt(curl, CURLOPT_SOCKOPTFUNCTION, sockopt_callback);
         curl_easy_setopt(curl, CURLOPT_SOCKOPTDATA, (void*)req);
+        curl_easy_setopt(curl, CURLOPT_FAILONERROR, 1L);
```

Comment:
According to cUrl documentation, this option "_tells the library to fail the request if the HTTP code returned is equal to or larger than 400._", while the default behavior would be to return the page as usual, ignoring the error.

I wonder if enabling the option would have any side effect on the management of `HttpReqXfer::status`. When it's set to `REQ_FAILURE`, it handles the error codes 509, 429, 403, 404 and 503. My concern is whether these codepaths will still be visited with this option enabled?

Reply:
Maybe we need to handle, in case of failed requests, the `CURLE_HTTP_RETURNED_ERROR`
_When this option is used and an error is detected, it will cause the connection to get closed and CURLE_HTTP_RETURNED_ERROR is returned._

Reply:
Yes, this case is handled already in CurlHttpIO::multidoio().   It passes the CURLMSG_DONE test, and assigns CURLE_HTTP_RETURNED_ERROR to errorCode from msg->data.result.   After extracting the http code and storing it in our Req object, it also assigns REQ_FAILURE to the status member after the comment "check httpstatus and response length".   So, I think we are already covered for this case, and it's working already for these test cases, logging the REQ_FAILURE and so on.   I think, for the normal case when the headers didn't have the `Connection: close` line, the old code worked fine.  But, when the request fails (>=400) and the connection close is triggered like that, the way cURL works internally, it just doesn't store a record for the curl_multi_info_read() to report.

Reply:
Alright, it sounds good. Thanks for all the investigation, Matt.
Ready to merge, IMO ;)